plugin/formatting.py

```diff
@@ -56,11 +67,74 @@ def run(self, _):
                     "uri": filename_to_uri(self.view.file_name())
                 },
                 "range": Range.from_region(self.view, region).to_lsp(),
-                "options": {
-                    "tabSize": self.view.settings().get("tab_size", 4),
-                    "insertSpaces": True
-                }
+                "options": get_formatting_options(self.view)
             }
             client.send_request(Request.rangeFormatting(params),
                                 lambda response: self.view.run_command('lsp_apply_document_edit',
                                                                        {'changes': response}))
+
+
+class DocumentOnTypeFormattingListener(sublime_plugin.ViewEventListener):
+
+    @classmethod
+    def is_applicable(cls, subl_settings: sublime.Settings) -> bool:
+        syntax = subl_settings.get('syntax')
+        if not syntax:
+            return False
+        elif not is_supported_syntax(syntax):
+            return False
+        return True
+
+    def __init__(self, view: sublime.View) -> None:
+        super().__init__(view)
+        self._triggers = []  # type: List[str]
+        self._client = None  # type: Optional[Client]
+        self._initialized = False
+
+    def on_modified_async(self) -> None:
+        if not settings.format_on_type:
+            return
+        if not self._ensure_initialized():
+            return
+        for region in self.view.sel():
+            self._on_type_formatting_async(region.begin())
+
+    def _ensure_initialized(self) -> bool:
+        if self._initialized:
+            return bool(self._client)
+        self._initialized = True
+        self._client = client_for_view(self.view)
+        if self._client:
+            options = self._client.get_capability("documentOnTypeFormattingProvider")
+            if options:
+                self._triggers = [options.get("firstTriggerCharacter")] or []
+                self._triggers.extend(options.get("moreTriggerCharacter") or [])
+                return True
+            else:
+                self._client = None
+                return False
+        else:
+            return False
+
+    def _on_type_formatting_async(self, point: int) -> None:
+        previous_character = self.view.substr(point - 1)
+        if previous_character not in self._triggers:
+            return
+        params = get_document_position(self.view, point)
+        if not params:
+            return
+        params["ch"] = previous_character
+        params["options"] = get_formatting_options(self.view)
+        purge_did_change(self.view.buffer_id())
```

Comment:
Typing `}}` here might cause a similar experience with repeated requests from typing `...` for the CompletionHandler. With completions we try to not block and discard old results, we might need something similar here.

Reply:
This ViewEventListener uses `on_modified_async`, so nothing will be executed on Sublime's UI/main thread.

I'm not sure how to solve your concern at this point.

Reply:
We don't need to solve it immediately :) Thanks for the updates, I'll look at merging this weekend!